### DISCUSSION TITLE
Changing globals in toolbar reloads whole site and sometimes don't work

### DISCUSSION THREAD

**Discussion by JuhG** [storybookjs/storybook#26941] [Help]
### Summary

Maybe these are two separate issues
- whenever we change the globals in the toolbar, the whole documentation page reloads, not just the stories (is it because the whole page is wrapped inside `ThemeProvider`?)
- in some cases it fails, doesn't update the globals and just reloads the old version. copied the error message and a video of it in "Additional information"

Anyone ran into something similar?

### Additional information

Here's a video of it:
https://github.com/storybookjs/storybook/assets/10765158/2d28e389-36e5-48a9-ac95-a69cc37da7b4

And here's the error message:
![Screenshot 2024-04-24 at 13 53 23](https://github.com/storybookjs/storybook/assets/10765158/25a4f132-2d72-4736-bf0b-00ac0ad6b4ba)

The error message in text:
`manager - received storyRenderPhaseChanged but was unable to determine the source of the event`
`manager - received globalsUpdated but was unable to determine the source of the event`

And here's the relevant part of the `preview.tsx`:
```tsx
const preview: Preview = {
  ...
  globalTypes: {
    mode: {
      defaultValue: "dark",
      toolbar: {
        title: "Mode",
        items: ["light", "dark"],
        dynamicTitle: true,
      },
    },
    ...
  },
  decorators: [
    (Story, context) => {
      return (
        <ThemeProvider ...>
          <Story />
        </ThemeProvider>
      );
    },
  ],
}
```

**Comment by JuhG** [comment]
Found a related issue: https://github.com/storybookjs/storybook/issues/15659
I'm going to expand on both of these issues and why they are hard to reproduce

### Changing globals reloads the documentation page

This is related to`preview.parameters.docs.container` or `decorators` - when you wrap the documentation page with `ThemeProvider` that uses these globals, it reloads the page, not just rerenders the component. For some reason it doesn't do this in a stackblitz example - tried to remove components and simplify the `main.stories` array and at some point it fixed itself. So looks like SB only does this when there are a lot of components, maybe some race condition or performance bottleneck we're hitting. But this is just a guess. We have something like 50+ components, so nothing crazy.

### Changing globals fails in some cases

`received X but was unable to determine the source of the event` - seems like `event.source` is undefined in `getEventSourceUrl`:
https://github.com/storybookjs/storybook/blob/38dfc50b7a4e2d4d1e59356362a461b1aa18c3cd/code/lib/channels/src/postmessage/index.ts#L218-L219
I'm not familiar with the code, so not sure why we need different logic on docs page and preview page. `rawEvent.origin` is always defined, but in that function the code looks for the source, which is not available sometimes. The strange part is that in some cases it works. A couple of components always work, others mostly fail, some fails in a way that changing the global doesn't even happen - which is the most frustrating part of this. (again, some of this is guessing)

So if I change that like to
```tsx
event.source = rawEvent.origin;
```
this issue goes away. Most probably there's a reason this needs to happen, but want to say that the issue is in `getEventSourceUrl.ts` I think.

**Comment by nukemonk** [comment]
Hello.

I've ran into something similar. The same issue happens to me in docs pages with "larger" stories. It indeed seems to be some race condition since the issue happens more frequently when stories grow bigger. It also happens more frequently in development than in production builds.
Another interesting thing is that it works fine when you change globals using the `useGlobals` hook in a decorator.

I've made a [reproduction repo](https://github.com/nukemonk/globals_toolbar_error) and hosted it with Chromatic [here](https://66ac97cfb6985b19d709445f-szcqjhqcuf.chromatic.com/?path=/docs/example-button-with-problem--docs). I have use Material-UI but I don't think that is the actual cause. It seems to happen without MUI as well as long the render phase of the story is long/heavy.
Just note that Chromatic hosts a productional build and therefor you don't see the error in the console and neither does it happen as frequently.

**Comment by JuhG** [comment]
Thanks @nukemonk, awesome example. Referenced this in the related issue.

This is how console looks like locally running it:
![Screenshot 2024-08-02 at 10 49 51](https://github.com/user-attachments/assets/b03617a0-392d-4af1-97b1-41b11be5f1fd)

**Comment by KurtRogers** [comment]
Any news on this? Experiencing it on `8.1.11`

**Reply by nukemonk** [reply]
Doesn't seem like it, unfortunately

**Comment by ChristopherJohnson25** [comment]
I am having the same issue on 8.6.0

**Comment by nukemonk** [comment]
The issue still persists on Storybook 9.1.8. The only difference is that at least the selected value persists. So it does not sometimes use the old value anymore.

**Reply by damisparks** [reply]
@JuhG or anyone managed to solve it?

**Reply by damisparks** [reply]
Never mind. I figured it out.

**Reply by JuhG** [reply]
hey @damisparks would you mind sharing what you've found?

**Reply by binaryfire** [reply]
Would also love to know if there's a workaround. Hitting the same problem with Storybook 10 and Tanstack Start

**Comment by volodymyr-leskiv** [comment]
The same issue on 10.2.6
